### PR TITLE
Improve `ora.promise()`

### LINKED ISSUE #169: Suggestion: promise enhancements
I often want to display different text based on the result of the `promise` (whether success or failure). Sometimes my `action` is multiple promises, so it would be convenient if the signature for that was `() => Promise` (so that I don't need to do create a new named variable for my promise on the line before it, and then pass it to ora). Lastly, it seems a lot more convenient to return the promise, rather than the spinner instance.

I've been using this little wrapper
```js
  const oraPromise = async <T>(
    action: () => Promise<T>,
    options: {
      startText?: () => string;
      successText?: (resp: T) => string;
      failText?: (err: Error) => string;
    } = {}
  ): Promise<T> => {
    const spinner = ora();
    const { startText, successText, failText } = options;
    if (startText) spinner.start(startText());
    return action()
      .then((resp) => {
        if (successText) spinner.succeed(successText(resp));
        return resp;
      })
      .catch((err) => {
        if (failText) spinner.fail(failText(err));
        throw err;
      });
  };
```
I think it enhances the usability of `promise` for some common cases. If you think it's worthwhile, I may be able to find time to submit a PR.

example:
```js
  // before
  await Promise.all(
    actions.map((name) => {
      const myAction = doSomeAsyncThing(name);
      const spinner = ora.promise(myAction, `Performing async action ${name}...`);
      myAction
        .then((result) => {
          spinner.succeed(`Success! ${result}`);
        })
        .catch((err) => {
          spinner.fail(`Error ${err}`);
        });
      return myAction;
    })
  );
```
```js
  // after
  await Promise.all(
    actions.map((name) => {
      return oraPromise(() => doSomeAsyncThing(name), {
        startText: () => `Performing async action ${name}...`,
        successText: (result) => `Success! ${result}`,
        failText: (err) => `Error ${err}`,
      });
    })
  );
```

### PR DESCRIPTION
**Pull request by SrBrahma**
Implemented https://github.com/sindresorhus/ora/issues/169#issuecomment-873269524

Improves ora.promise:
1) The action can now be a function. It will automatically be executed. The corresponding spinner will be used as argument to that function.
2) It will now return a promise, that resolves when ora.promise given promise/function is resolved, so the user can `await ora.promise(..)`.
3) Added successText and failText, that will be used in the spinner.succeed()/spinner.fail(). They can be either a string or a function. Being a function, successText will receive as argument the return of the action resolve, and failText will receive the action reject content.

Fixes #169

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -21,23 +21,23 @@ declare namespace ora {
 
 	interface Options {
 		/**
-		Text to display after the spinner.
-		*/
+		 Text to display after the spinner.
+		 */
```
Comment (sindresorhus) [inline index.d.ts:25]: Don't do unrelated changes.

File: index.d.ts [hunk 2]
```diff
@@ -262,16 +275,16 @@ declare const ora: {
 	(options?: ora.Options | string): ora.Ora;
 
 	/**
-	Starts a spinner for a promise. The spinner is stopped with `.succeed()` if the promise fulfills or with `.fail()` if it rejects.
+	Starts a spinner for a function or a promise. The spinner is stopped with `.succeed()` if the promise fulfills or with `.fail()` if it rejects.
```
Comment (sindresorhus) [inline index.d.ts:278]: index.d.ts and readme should be in sync.

**Comment by sindresorhus** [comment]
The pull request needs a proper descriptive title.

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "ora",
-	"version": "5.4.1",
+	"version": "6.0.0",
```
Comment (sindresorhus) [inline package.json:3]: This should be done by the maintainers when they're ready to release, not in a PR.

**Comment by sindresorhus** [comment]
Please always run `npm test` locally before pushing changes.

File: index.d.ts (hunk 1 above)
Comment (SrBrahma) [reply index.d.ts:25]: Sorry, I think that happened when I changed from spaces to tab in my vscode.

**Comment by SrBrahma** [comment]
> Please always run `npm test` locally before pushing changes.

I did and fixed my own errors, but there were other errors not related to my changes

**Comment by SrBrahma** [comment]
I have no idea on how to fix this npm test issue:
image

**Comment by SrBrahma** [comment]
I found out why the test error was happening. The new promise() was calling the promise that always rejected. I thought that the promise was being rejected before entering the promise function call. Also, was the PromiseOptions type definition was missing, looks like I only had added it to the README.

File: test.js [hunk 4]
```diff
@@ -167,16 +167,12 @@ test('.promise() - rejects', async t => {
...
 		stream,
 		text: 'foo',
 		color: false,
 		isEnabled: true
-	});
-
-	try {
-		await rejects;
-	} catch {}
+	}).catch(() => null);
```
Comment (sindresorhus) [inline test.js:175]: Use try/catch instead of `.catch()`

File: index.d.ts [hunk 5]
```diff
@@ -133,6 +133,17 @@ declare namespace ora {
...
 	}
 
+	interface PromiseOptions<T> extends Options {
+		/** The new text of the spinner when the promise is resolved.
+     *
+     * * If undefined, will keep the initial text. */
+		successText?: ((resp: T) => string) | string;
+		/** The new text of the spinner when the promise is rejected.
+     *
+     * If undefined, will keep the initial text. */
```
Comment (sindresorhus) [inline index.d.ts:143]: This needs to be improved. Incorrect indentation and the formatting looks messy.

File: index.d.ts [hunk 6]
```diff
@@ -133,6 +133,17 @@ declare namespace ora {
 		readonly prefixText?: string | PrefixTextGenerator;
 	}
 
+	interface PromiseOptions<T> extends Options {
+		/** The new text of the spinner when the promise is resolved.
+     *
+     * * If undefined, will keep the initial text. */
+		successText?: ((resp: T) => string) | string;
```
Comment (sindresorhus) [inline index.d.ts:140]: Don't use abbreviations.

File: readme.md [hunk 7]
```diff
@@ -226,11 +226,26 @@ Change the spinner indent.
...
+ora.Options & {
+/** The new text of the spinner when the promise is resolved.
+
+If undefined, will keep the initial text. */
+successText?: ((resp: T) => string) | string;
+/** The new text of the spinner when the promise is rejected.
+
+If undefined, will keep the initial text. */
+failText?: ((error: Error) => string) | string;
+```
```
Comment (sindresorhus) [inline readme.md:248]: The readme docs should use the same formatting as used here: https://github.com/sindresorhus/boxen#options (Not just a copy paste of the TS types).

File: index.d.ts (hunk 5 above)
Comment (SrBrahma) [reply index.d.ts:143]: Hmmm, I haven't looked at the result. After all the xo's errors, I just `xo --fix` and believed it would fix everything properly. I will soon do your requests.

File: readme.md [hunk 8]
```diff
@@ -226,11 +226,28 @@ Change the spinner indent.
...
+Type: `Promise | ((spinner: ora.Ora) => Promise)`
+
+#### options
+
+Type: `object`
+
+All of the [options](#options) plus the following:
+
+##### successText
+Type: `string | ((result: T) => string)`
```
Comment (SrBrahma) [inline readme.md:242]: For the functions, I used this as the pattern: https://github.com/sindresorhus/got/blob/main/documentation/5-https.md#checkserveridentity. Please, change it further if wanted to fit your README styling.

File: index.d.ts [hunk 9]
```diff
@@ -133,6 +133,22 @@ declare namespace ora {
...
+    If undefined, will keep the initial text.
+    */
+		successText?: string | ((result: T) => string);
+
+		/**
+    The new text of the spinner when the promise is rejected.
+
+    If undefined, will keep the initial text.
+    */
+		failText?: string | ((error: Error) => string);
```
Comment (sindresorhus) [inline index.d.ts:149]: Still messy formatting. Please look over the whole diff.

Comment (SrBrahma) [reply index.d.ts:149]: I don't know why this is happening at all. In my VSCode the indentation is right

Comment (SrBrahma) [reply index.d.ts]: I didn't have EditorConfig extension installed here, I was using 2 spaces instead of tab (size 2). The code was looking the same in VSCode, but looks like the diff in GitHub renders tabs as 8 spaces, that's why it was visually strange.

I activated the whitespace rendering, and that's what was happening:

image

I found out that you are the XO dev. Maybe would be a good idea to have a pre-commit/GitHub action to lint this. Or even, to throw a warn/error on test if spaces were found instead of tabs. In eslint, this is doable.

I hope everything is fixed now.

Comment (SrBrahma) [reply index.d.ts]: Just for your information, there is a cool eslint plugin for JSDoc: https://github.com/gajus/eslint-plugin-jsdoc

I will soon add it to my personal eslint-config, as I also have my personal taste for JSDoc comments. Maybe you could implement a similar implementation for XO to avoid those issues among all your repos, in PR like mine.

Comment (sindresorhus) [reply index.d.ts]: > Or even, to throw a warn/error on test if spaces were found instead of tabs. In eslint, this is doable.

That's already a thing. Ora just didn't have the latest XO version.

However, you are still expected to look over your own PR diff.

Comment (sindresorhus) [reply index.d.ts]: > Just for your information, there is a cool eslint plugin for JSDoc: gajus/eslint-plugin-jsdoc

I plan to add it at some point, but it's some work to evaluate all the rules, and I haven't had the time to do that yet: https://github.com/xojs/xo/issues/378

**Comment by sindresorhus** [comment]
Thanks :)
